# Objective:

# Contents:

# Data Description:

# Importing necessary libraries, Loading Data and Overview

In [1]:
# To help with reading and manipulating data
import pandas as pd
import numpy as np

# To help with data visualization
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns

# To be used for missing value imputation
from sklearn.impute import SimpleImputer

# To help with model building
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import (
    AdaBoostClassifier,
    GradientBoostingClassifier,
    RandomForestClassifier,
    BaggingClassifier,
)

# To get different metric scores, and split data
from sklearn import metrics
from sklearn.model_selection import train_test_split, StratifiedKFold, cross_val_score
from sklearn.metrics import (
    f1_score,
    accuracy_score,
    recall_score,
    precision_score,
    confusion_matrix,
    roc_auc_score,
    plot_confusion_matrix,
)

# To be used for data scaling and one hot encoding
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder

# To be used for tuning the model
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

# To be used for creating pipelines and personalizing them
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

# To define maximum number of columns to be displayed in a dataframe
pd.set_option("display.max_columns", None)

# To supress scientific notations for a dataframe
pd.set_option("display.float_format", lambda x: "%.3f" % x)

# To supress warnings
import warnings

warnings.filterwarnings("ignore")

# This will help in making the Python code more structured automatically (good coding practice)
#%load_ext nb_black

# Loading the dataset

In [2]:
df = pd.read_csv("hotel_bookings 2.csv")

In [3]:
df

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  children  babies meal country  \
0                          0       2     0.000       0   BB     PRT   
1                          0       2     0.000       0   BB     PRT   
2                          1       1     0.000       0   BB     GBR   
3                          1       1     0.000       0   BB     GBR   
4                          2       2     0.000       0   BB     GBR   
...                      ...     ...       ...     ...  ...     ...   
119385                     5       2     0.000       0   BB     BEL   
119386                     5       3     0.000       0   BB     FRA   
119387                     5       2     0.000       0   BB     DEU   
119388                     5       2     0.000       0   BB     GBR   
119389                     7       2     0.000       0   HB     DEU   

       market_segment distribution_channel  is_repeated_guest  \
0              Direct               Direct                  0   
1              Direct               Direct                  0   
2              Direct               Direct                  0   
3           Corporate            Corporate                  0   
4           Online TA                TA/TO                  0   
...               ...                  ...                ...   
119385  Offline TA/TO                TA/TO                  0   
119386      Online TA                TA/TO                  0   
119387      Online TA                TA/TO                  0   
119388      Online TA                TA/TO                  0   
119389      Online TA                TA/TO                  0   

        previous_cancellations  previous_bookings_not_canceled  \
0                            0                               0   
1                            0        

# Data Overview

The initial steps to get an overview of any dataset is to:

observe the first few rows of the dataset, to check whether the dataset has been loaded properly or not
get information about the number of rows and columns in the dataset
find out the data types of the columns to ensure that data is stored in the preferred format and the value of each property is as expected.
check the statistical summary of the dataset to get an overview of the numerical columns of the data

# Displaying the first few rows of the dataset

In [4]:
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2     0.000       0   
1                        0                     0       2     0.000       0   
2                        0                     1       1     0.000       0   
3                        0                     1       1     0.000       0   
4                        0                     2       2     0.000       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     GBR         Direct               Direct                  0   
3   BB     GBR      Corporate            Corporate                  0   
4   BB     GBR      Online TA                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled reserved_room_type  \
0                       0                               0                  C   
1                       0                               0                  C   
2                       0                               0                  A   
3                       0                               0                  A   
4                       0                               0                  A   

  assigned_room_type  booking_changes deposit_type   agent  company  \
0                  C                3   No Deposit     NaN      NaN   
1                  C                4   No Deposit     NaN      NaN   
2                  C                0   No Deposit     NaN      NaN   
3                  A                0   No Deposit 304.000      NaN   
4                  A                0   No Deposit 240.000      NaN   

   days_in_waiting_list customer_type    adr  required_car_parking_spaces  \
0                     0     Transient  0.000                            0   
1                     0     Transient  0.000                            0   
2                     0     Transient 75.000                            0   
3                     0     Transient 75.000                            0   
4                     0     Transient 98.000                            0   

   total_of_special_requests reservation_status reservation_status_date  
0                          0          Check-Out                1/7/2015  
1                          0          Check-Out                1/7/2015  
2                          0          Check-Out                2/7/2015  
3                          0          Check-Out                2/7/2015  
4                          1          Check-Out                3/7/2015

In [6]:
df.tail(10)

hotel  is_canceled  lead_time  arrival_date_year  \
119380  City Hotel            0         44               2017   
119381  City Hotel            0        188               2017   
119382  City Hotel            0        135               2017   
119383  City Hotel            0        164               2017   
119384  City Hotel            0         21               2017   
119385  City Hotel            0         23               2017   
119386  City Hotel            0        102               2017   
119387  City Hotel            0         34               2017   
119388  City Hotel            0        109               2017   
119389  City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
119380             August                        35   
119381             August                        35   
119382             August                        35   
119383             August                        35   
119384             August                        35   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
119380                         31                        1   
119381                         31                        2   
119382                         30                        2   
119383                         31                        2   
119384                         30                        2   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  children  babies meal country  \
119380                     3       2     0.000       0   SC     DEU   
119381                     3       2     0.000       0   BB     DEU   
119382                     4       3     0.000       0   BB     JPN   
119383                     4       2     0.000       0   BB     DEU   
119384                     5       2     0.000       0   BB     BEL   
119385                     5       2     0.000       0   BB     BEL   
119386                     5       3     0.000       0   BB     FRA   
119387                     5       2     0.000       0   BB     DEU   
119388                     5       2     0.000       0   BB     GBR   
119389                     7       2     0.000       0   HB     DEU   

       market_segment distribution_channel  is_repeated_guest  \
119380      Online TA                TA/TO                  0   
119381         Direct               Direct                  0   
119382      Online TA                TA/TO                  0   
119383  Offline TA/TO                TA/TO                  0   
119384  Offline TA/TO                TA/TO                  0   
119385  Offline TA/TO                TA/TO                  0   
119386      Online TA                TA/TO                  0   
119387      Online TA                TA/TO                  0   
119388      Online TA                TA/TO                  0   
119389      Online TA                TA/TO                  0   

        previous_cancellations  previous_bookings_not_canceled  \
119380                       0                               0   
119381                       0                               0   
119382                       0                               0   
119383                       0                               0   
119384                       0                               0   
119385                       0                               0   
119386                       0                  

# Checking the shape of the dataset

In [7]:
# Checking the number of rows and columns in the data
df.shape

(119390, 32)

The dataset contains information (32 attributes) about 119390 rows

# Checking the data types of the columns for the dataset

In [8]:
# let's check the data types of the columns in the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

There are a total of 32 columns and 2,240 observations in the dataset
We can see that children,country, agent and company columns have null values
reservation_status_date should be date type

In [12]:
df['reservation_status_date']=pd.to_datetime(df['reservation_status_date'])

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   hotel                           119390 non-null  object        
 1   is_canceled                     119390 non-null  int64         
 2   lead_time                       119390 non-null  int64         
 3   arrival_date_year               119390 non-null  int64         
 4   arrival_date_month              119390 non-null  object        
 5   arrival_date_week_number        119390 non-null  int64         
 6   arrival_date_day_of_month       119390 non-null  int64         
 7   stays_in_weekend_nights         119390 non-null  int64         
 8   stays_in_week_nights            119390 non-null  int64         
 9   adults                          119390 non-null  int64         
 10  children                        119386 non-null  float64

Successfully changing the reservation_status_date variable data type .

# Statistical summary of the dataset

In [14]:
df.describe(include="object").T

count unique         top    freq
hotel                 119390      2  City Hotel   79330
arrival_date_month    119390     12      August   13877
meal                  119390      5          BB   92310
country               118902    177         PRT   48590
market_segment        119390      8   Online TA   56477
distribution_channel  119390      5       TA/TO   97870
reserved_room_type    119390     10           A   85994
assigned_room_type    119390     12           A   74053
deposit_type          119390      3  No Deposit  104641
customer_type         119390      4   Transient   89613
reservation_status    119390      3   Check-Out   75166

In [16]:
for col in df.describe(include="object").columns:
    print (col)
    print (df[col].unique())
    print ('_'*50)

hotel
['Resort Hotel' 'City Hotel']
__________________________________________________
arrival_date_month
['July' 'August' 'September' 'October' 'November' 'December' 'January'
 'February' 'March' 'April' 'May' 'June']
__________________________________________________
meal
['BB' 'FB' 'HB' 'SC' 'Undefined']
__________________________________________________
country
['PRT' 'GBR' 'USA' 'ESP' 'IRL' 'FRA' nan 'ROU' 'NOR' 'OMN' 'ARG' 'POL'
 'DEU' 'BEL' 'CHE' 'CN' 'GRC' 'ITA' 'NLD' 'DNK' 'RUS' 'SWE' 'AUS' 'EST'
 'CZE' 'BRA' 'FIN' 'MOZ' 'BWA' 'LUX' 'SVN' 'ALB' 'IND' 'CHN' 'MEX' 'MAR'
 'UKR' 'SMR' 'LVA' 'PRI' 'SRB' 'CHL' 'AUT' 'BLR' 'LTU' 'TUR' 'ZAF' 'AGO'
 'ISR' 'CYM' 'ZMB' 'CPV' 'ZWE' 'DZA' 'KOR' 'CRI' 'HUN' 'ARE' 'TUN' 'JAM'
 'HRV' 'HKG' 'IRN' 'GEO' 'AND' 'GIB' 'URY' 'JEY' 'CAF' 'CYP' 'COL' 'GGY'
 'KWT' 'NGA' 'MDV' 'VEN' 'SVK' 'FJI' 'KAZ' 'PAK' 'IDN' 'LBN' 'PHL' 'SEN'
 'SYC' 'AZE' 'BHR' 'NZL' 'THA' 'DOM' 'MKD' 'MYS' 'ARM' 'JPN' 'LKA' 'CUB'
 'CMR' 'BIH' 'MUS' 'COM' 'SUR' 'UGA' 'BGR' 'CIV' 'J

# Observations:
Year_Birth has a large range of values i.e. 1893 to 1996. The birth year <=1900 is some users.. It's very unlikely that the person is alive. It may be a reporting error. We will explore this further. Income has a max value of 666666 which is far greater than the mean and could be an outlier. Columns - Income, MntFruits, MntWines, MntMeatProducts, MntFishProducts, MntSweetProducts might have outliers on the right end as there is a large difference between 75th percentile and maximum values. Recency has an approx equal mean and median which is equal to 49 days. Highest mean amount spent per customer in the last two years is on wines (approx 304), followed by meat products (approx 167).

# Checking for duplicate values

In [10]:
df.duplicated().sum()

31994

In [22]:
# let's check for missing values in the data
df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [24]:
df.drop(['company','agent'], axis= 1, inplace= True)
df.dropna (inplace= True)

In [25]:
# let's check for missing values in the data
df.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests   